### DISCUSSION TITLE
Tupple support in linq

### DISCUSSION THREAD

**Discussion by mariodvm** [dotnet/csharplang#5386] [Language Ideas]
Suppose you want a bounding box of many points (x,y) on a data base. The SQL is like this:
`SELECT MIN(X), MIN(Y), MAX(X), MAX(Y)
FROM MYTABLE`
This cannot be generated as a single query by a LinqQueryProvider because something like tupples is required to be supported in the expresión tree. It could be written like this:
`var q = myTable.Select (t => (myTable.Min(s => s.X), myTable.Min(s => s.Y), myTable.Max(s => s.X), myTable.Max(s => s.Y)))`

And this is the proposals: To include tupples in expresión trees in order to make these linq queries possible

**Comment by HaloFour** [comment]
You can use anonymous classes for intermediate projections within LINQ queries.

**Reply by jnm2** [reply]
Modifying the given code sample:

```cs
var q = myTable.Select(t => new { MinX = Math.Min(t.X), MinY = Math.Min(t.Y), MaxX = Math.Max(t.X), MaxY = Math.Max(t.Y) });
```

**Reply by mariodvm** [reply]
yes, but the actual Min, and Max methods needed are not those of the Math class but the aggregate extensions declared on Queryable, but they must be used once in a single query chain. The goal is to make it possible to get that final SQL from Linq somehow. I think that tupples could be good, even when the solution could require additional features

**Reply by mariodvm** [reply]
I updated the original example to make it clearer by using the aggregate methods Min and Max, but this is not the solution, because the queryable may be a complex expression, not always a wellknown variable (_myTable_ in this case)

**Reply by HaloFour** [reply]
It sounds like what you're asking for are the grouping and aggregation capabilities of LINQ which do already exist and are unrelated to tuples:

```cs
var q = myTable
    .GroupBy(t => true) // using a sentinel value to create a single grouping
    .Select(group => new {
        MinX = group.Min(t => t.X),
        MinY = group.Min(t => t.Y),
        MaxX = group.Max(t => t.X),
        MaxY = group.Max(t => t.Y)
    });
```

**Comment by MgSam** [comment]
There are TONS of SQL constructs that can not be created via LinqToSql / Entity Framework. The impedance mismatch between imperative languages and SQL is quite high. For anything but simple queries with simple joins, you're better off making a view or proc in SQL or sending a hand-written query to the database through ADO.NET. Aggregation in particular is not something I'd try to leave to LinqToSql.

**Comment by singh-rajiv** [comment]
### Automatic deconstruction of a tuple in LINQ on a sequence of tuples

I have a code snippet like below which is a projection of a sequence of tuples. However, I noticed that instead of automatically deconstructing the tuples in (key, val) in the Select call, the compiler binds it to the Select overload that takes in integer index of the element. So the inferred type of "key" in the Select method is (string, string) and that of "val" is int.

```csharp
        var val1 = "somestring1";
        var val2 = "somestring2";
        var val3 = "somestring3";
        var encodedKeyValuePairs = 
            new [] 
            { 
                ("key1", val1), 
                ("key2", val2), 
                ("key3", val3)
            }.
            Select((key, val) => $"{HttpUtility.UrlEncode(key)}={HttpUtility.UrlEncode(val)}");
```
It'd be nice if the compiler would automatically deconstruct the tuple into individual items with the correct type. I can foresee the challenge where the type of the second item in the tuple is also int.

But that could be handled by writing something like below if I want to use index along with the deconstructed tuple.
```csharp
Select(((key, val), idx)) => $"{key}={val} & index={idx}")
```

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/discussions/258

Of course the problem is that `Select((key, val) => ...)` is already legal and means something completely different so it would not be possible to deconstruct the tuples using that syntax.
